### PR TITLE
Add methods getPeriodList and getTemporalList #503

### LINKED ISSUE #503: Missing methods: `getPeriodList` and `getTemporalList`
All methods `getX` in Typesafe config have an equivalent `getXList` method. Methods `getPeriod` and `getTemporal`, introduced by #478, lack their "list" equivalents.

### PR DESCRIPTION
**Pull request by marconilanna**
@ktoso @kag0 fixes #503

### CODE REVIEW

**Comment by marconilanna** [comment]
bump (all checks passed)

**Comment by gmethvin** [comment]
I still don't agree with having a `getTemporal` or `getTemporalList` but I guess it's too late to object now...

**Comment by ktoso** [comment]
Why "still" was this discussed somewhere? Please provide context @gmethvin; I've only seen the ticket over there

File: config/src/main/java/com/typesafe/config/impl/SimpleConfig.java [hunk 1]
```diff
@@ -545,6 +545,50 @@ public TemporalAmount getTemporal(String path){
...
+            if (v.valueType() == ConfigValueType.STRING) {
+                value = (String) v.unwrapped();
+            } else if (v.valueType() == ConfigValueType.NUMBER) {
+                value = "" + ((Number) v.unwrapped()).longValue();
+            } else {
+                throw new ConfigException.WrongType(v.origin(), path,
+                        "duration or period string",
+                        v.valueType().name());
+            }
+            try{
```
Comment (gmethvin) [inline config/src/main/java/com/typesafe/config/impl/SimpleConfig.java]: Can you just factor out the logic from `getTemporal` into a method?

**Comment by ktoso** [comment]
Hm, I see in the other thread; will read that

**Comment by gmethvin** [comment]
@ktoso `TemporalAmount` is not really meant to be used by end users, and it's not trivial to convert to an actual useful value. Something like `Duration.from(temporalAmount)` will throw an exception for `Period`s.

**Comment by marconilanna** [comment]
@gmethvin @ktoso Have we reached a conclusion regarding `getTemporal` and `getTemporalList`?

I'm not advocating in favor or against them but, for consistency, we should either have both or have neither.

**Comment by marconilanna** [comment]
@ktoso @gmethvin bump

**Comment by ktoso** [comment]
No strong opinion, WDYT @havocp ?

**Comment by havocp** [comment]
I havent had a chance to dig in, I’m not familiar with temporal and period so would need to read the docs. Also of course we can’t remove API we already shipped, if we did ship it.

**Comment by marconilanna** [comment]
@havocp We could deprecate `getTemporal` and remove `getTemporalList` from this PR.

**Comment by gmethvin** [comment]
I'm fine with leaving in the API we already shipped, but not sure what the use case is for adding these new `getTemporal` methods.

`Period` is a length of time in the ISO-8601 calendar (e.g. 1 year 2 months 3 days), whereas `Duration` is an exact length of time (e.g. 1 million seconds). Usually you want one or the other, and in most config use cases you'd probably want a `Duration`.

I suspect the original use case for `getTemporal` was so you could have a length of time like `1 year` rather than `365 days` (this is what the [original issue](https://github.com/lightbend/config/issues/473) suggests). Unfortunately it's not trivial to convert from a `TemporalAmount` to an exact length of time. I would much prefer to have `getDuration` and other duration methods support units like `years` and `months` and convert them using the standard `ChronoUnit` amounts (e.g. 1 year = 365 days). If someone really wants a `Period` they can use `getPeriod`.

Anyway, I don't really feel *that* strongly on this as it's pretty easy to just avoid using `getTemporal` in the API. I'm just explaining what I think the right API design is.

**Comment by marconilanna** [comment]
maybe @kag0 could explain his motivation?

**Comment by kag0** [comment]
Hey all,
I think @gmethvin is right in that `Temporal` is a tricky abstraction between `Duration` and `Period`. 
Imo having access to a `Temporal` generically could be useful to framework creators at a low level (when something like `TemporalUnit.between` is used), but in the majority of cases it's better to have one or the other explicitly. 
I don't have a real opinion on if `getTemporalList` should be included publicly, but as I see it it's api consistency vs the increasingly rare usage of the method. Also it's always easier to make it private now and switch to public if users request, than it is to make it public now and switch to private later if issues arise.

**Comment by marconilanna** [comment]
Is there still interest in this PR?

**Comment by kdubb** [comment]
@marconilanna @kag0 Why was this just closed? Regardless of the discussion around the temporal API, it exists and should have consistency with other API (unless its deprecated, which as of now it is not) and have `getXXXList` methods. Also, `getPeriodList` was not under discussion and it was never added either.

**Pull request closed without merging**